### PR TITLE
ARROW-3399: [Python] Implementing numpy matrix serialization

### PR DESCRIPTION
**Pull request by rok**
See [ARROW-3399](https://issues.apache.org/jira/browse/ARROW-3399).

### CODE REVIEW

**Comment by mitar** [comment]
Are you sure this works (I mean it looks it does, it is a rhetorical question)? I have no idea why then I was getting an error when I tried this. I must have made some mistake.

This is great.

File: python/pyarrow/tests/test_plasma.py [hunk 1]
```diff
@@ -390,6 +390,12 @@ def __init__(self, val):
             object_id, -1, serialization_context=serialization_context)
         assert result.val == val.val
 
+    def test_store_np_matrix(self):
```
Comment (robertnishihara) [inline python/pyarrow/tests/test_plasma.py]: Can we test a few more dtypes as well?

File: python/pyarrow/tests/test_plasma.py [hunk 2]
```diff
@@ -390,6 +390,12 @@ def __init__(self, val):
...
+        m = np.matrix(np.random.randn(10, 4))
+        object_id = self.plasma_client.put(m)
+        buf = self.plasma_client.get(object_id)
+        assert (buf == m).all()
```
Comment (robertnishihara) [inline python/pyarrow/tests/test_plasma.py]: Will this catch errors if the dtype changes?

File: python/pyarrow/serialization.py [hunk 3]
```diff
@@ -55,6 +55,25 @@ def _deserialize_numpy_array_list(data):
...
+        # Make the array c_contiguous if necessary so that we can call change
+        # the view.
+        if not obj.flags.c_contiguous:
+            obj = np.ascontiguousarray(obj.A)
+        return obj.A.view('uint8'), obj.A.dtype.str
+    else:
+        return obj.A.tolist(), obj.A.dtype.str
+
+
+def _deserialize_numpy_matrix(data):
```
Comment (robertnishihara) [inline python/pyarrow/serialization.py]: I just did a little timing and the time taken by this method seems to scale linearly in the size of the array. Can we fix this?

I think `np.matrix(...)` is the slow line.

**Review by robertnishihara — COMMENTED** [review]
Thanks for doing this! Nice work.

File: python/pyarrow/tests/test_plasma.py (hunk 1 above)
Comment (rok) [reply python/pyarrow/tests/test_plasma.py]: I've added other types. Do you think `str`, `int` and `float` is enough?

File: python/pyarrow/tests/test_plasma.py (hunk 2 above)
Comment (rok) [reply python/pyarrow/tests/test_plasma.py]: It wouldn't e.g. for `int` and `float`. I've added a `dtype` comparison.

**Comment by rok** [comment]
I'm having issues building, any idea why after running:
```
pushd arrow/python
export PYARROW_WITH_FLIGHT=1
export PYARROW_WITH_GANDIVA=1
export PYARROW_WITH_ORC=1
export PYARROW_WITH_PARQUET=1
python setup.py build_ext --build-type=$ARROW_BUILD_TYPE --inplace
popd
```
I get:
```
...
[100%] Linking CXX shared module release/_parquet.cpython-37m-x86_64-linux-gnu.so
[100%] Built target _parquet
-- Finished cmake --build for pyarrow
Bundling includes: include
error: [Errno 2] No such file or directory: 'include'
```

I'm using conda and this worked up until some point recently and now stopped. Any ideas?

File: python/pyarrow/serialization.py (hunk 3 above)
Comment (rok) [reply python/pyarrow/serialization.py]: Indeed, copying doesn't seem like the thing to do here. I think `np.matrix(data, copy=False)` would just create a new view.

File: python/pyarrow/tests/test_plasma.py (hunk 1 above)
Comment (robertnishihara) [reply python/pyarrow/tests/test_plasma.py]: That's pretty good, but are we actually hitting the `if obj.dtype.str != '|O':` code path? Maybe add a test that uses a numpy matrix that contains some custom object, like

```python
class Foo(object):
    pass

x = np.matrix([Foo()])
```

There should be an analogous test for numpy arrays. If so, maybe copy that one.

**Comment by robertnishihara** [comment]
I'm not sure about the compilation error, but it's unrelated to this PR.

**Comment by rok** [comment]
I figured out my build issue. It was kinda stupid in hindsight. I opened an [issue on Jira](https://issues.apache.org/jira/projects/ARROW/issues/ARROW-5165).

File: python/pyarrow/tests/test_plasma.py (hunk 1 above)
Comment (rok) [reply python/pyarrow/tests/test_plasma.py]: I've added `CustomType` to the list of types we check for. I've added some logging locally and we appear to be hitting `if obj.dtype.str != '|O':` now.
We do get a lot of `np.matrix` deprecation warnings, should we silence them?

Comment (mitar) [reply python/pyarrow/tests/test_plasma.py]: > We do get a lot of np.matrix deprecation warnings, should we silence them?

In tests you mean?

Comment (rok) [reply python/pyarrow/tests/test_plasma.py]: Yes tests trigger them. I am not sure how it works if this is merged into arrow - would every deserialization trigger warnings or not?

Comment (mitar) [reply python/pyarrow/tests/test_plasma.py]: I think we could silence them during tests. But during regular use I would just leave to the user to do so if they want.

Comment (rok) [reply python/pyarrow/tests/test_plasma.py]: I don't have a strong opinion either way. What is the project default @robertnishihara?

**Comment by robertnishihara** [comment]
@rok nice work! I should have mentioned this earlier, but didn't notice until just now. This test should be in `test_serialization.py` instead of `test_plasma.py`. That will require the test to be changed a little bit (to not use plasma).

File: python/pyarrow/tests/test_plasma.py (hunk 1 above)
Comment (robertnishihara) [reply python/pyarrow/tests/test_plasma.py]: I don't see the warnings. Where are they? Did they appear in the arrow CI?

Comment (rok) [reply python/pyarrow/tests/test_plasma.py]: Odd - locally I see a warning similar to [this one](https://travis-ci.org/apache/arrow/jobs/519253981#L3149) about `np.matrix` being deprecated. I can't find  it in the CI log. I like @mitar's point I'll leave this as is.

**Comment by rok** [comment]
Thanks @robertnishihara :).
Yeah `test_serialization` does make more sense. I pushed a change now.

**Comment by robertnishihara** [comment]
Thanks @rok! Nice work, I merged this.

**Comment by rok** [comment]
Nice! Thanks @robertnishihara and @mitar!

**Pull request closed without merging**